text/0006-wasm-smart-contracts.md

```diff
@@ -0,0 +1,440 @@
+
+For example:
+
+```rust
+// macros for marking up the contract
+use contract_macros::contract_impl;
+use contract_macros::transaction;
+
+/// Structure for the AssetContract, on which implemenation transaction functions will be added
+pub struct AssetContract {
+
+}
+/// Implementation of the contract trait for the AssetContract
+/// There are default implementation methods, but can be modified if you wish
+/// 
+/// Recommended that the name() function is always modified
+impl Contract for AssetContract {
+    //! Name of the contract
+    fn name(&self) -> String {
+        format!("AssetContract")
+    }
+}
+
+#[contract_impl]
+impl AssetContract {
+    #[transaction]
+    pub fn asset_exists(assset_id:  String) -> Result<bool,ContractError> {
+        let ledger = Ledger::access_ledger();
+        let world = ledger.get_collection(CollectionName::World);
+
+        Ok(world.state_exists(assset_id))
+    }
+}
+```
+
+The developer could write several such contracts. As rust is a static language without runtime introspection, the developer will need to explicitly identify the contracts to the runtime.
+
+For example:
+
+```rust
+launch_handler!(register);
+
+/// Register the contracts
+/// Look to fold this into the macro (or the contract_impl macro)
+pub fn register() -> i32 {
+    ContractManager::register_contract(Box::new(AssetContract::new()));
+
+    // return 0 to indicate success, other values imply failure and will terminate
+    // the chaincode container
+    0
+}
+```
+
+### Ledger
+
+A Ledger's current value is held in a set of collections. A single ledger is available to any given deployed contract, though other ledgers may be available within the Network Channel and can be accessed via other contracts
+
+Of the collections that are reference by the ledger, there a single collection in the ledger that is public (the 'WorldState') and held by every organization in the Network Channel. A ledger may have 1 or more private collections that are available only in some organizations. Starting at Fabric v2.0 there will be an implicit private collection per organization.
+
+## Collection
+
+A collection is a set of states, with each state holding a business object or data. Each state being addressed by a key. Private collections are identified by name and held within a set of policy-defined organizations.
+
+Accessing any given collection is via the Ledger instance. 
+
+```rust
+    let Ledger = Ledger::access_ledger()
+
+    // predefined constant for the world state
+    let world = ledger.get_collection(CollectionName::World);
+
+    // Private Data collections are reference by name
+    let private_collection = ledger.get_collection(CollectionName::Private(String::from("PrivateSalesData")));
+
+    // Each organization will have an implicit private data collection created for them
+    let orgs_collection = ledger.get_collection(CollectionName::Organization(String::from("org1")));
```

Comment:
These two comments are applicable here too but I am fine with it if maintainers do not have any objections.  
https://github.com/hyperledger/fabric-rfcs/pull/16#discussion_r376212045 
https://github.com/hyperledger/fabric-rfcs/pull/16#discussion_r375970927